### DISCUSSION TITLE
How to access fprime html UI

### DISCUSSION THREAD

**Discussion by nchanilk** [nasa/fprime#1355] [General]
Hi,
I'm new to fprime, I've successfully installed fprime and ran the below command
fprime-gds -g html -r /root/fprime/Ref/build-artifacts. But I'm not getting any conntent when I do curl http://localhost:50050/50000. Is supposed to work for http? How can I test/validate the app? Any help would be appreciated.

Thanks,
Anil

**Comment by LeStarch** [comment]
The HTML GUI runs on a default port of 5000.  You should try to curl: `http://localhost:5000`.

The other two ports are for:

5000: user connects to ground system
50000: embedded software connects to ground system.
50050: internal data movement port for ground system

**Comment by nchanilk** [comment]
Thanks, @LeStarch  it worked. Is there a way I can change the localhost to my server IP?

Thanks,
Anil

**Reply by LeStarch** [reply]
--gui-addr GUI_ADDR   Set the GUI server address [default: 127.0.0.1]
  --gui-port GUI_PORT   Set the GUI server address [default: 5000]

These two arguments to fprime-gds.  You can set the address and port. To use whatever your server is, `--gui-addr 0.0.0.0`.  Note: you may need to expose the port through the server firewall to access it off the server itself.

Also, the GDS can control your application, so it is best to setup a security setup using a webserver as a proxy if you are working on a public network.

**Comment by nchanilk** [comment]
Thanks, @LeStarch  for your great and timely help, appreciated. 
I'm able to access the UI now, but it's not functioning due to the Bad Requests. (pelase find the below screen shot) 
I'm exploring the fPrime and eventually we will add more security layers and planning to integrate with the Cosmos.
In case if it's not too much ask, can you please help me /pointing to the resources on how uplink, downlink works from the backend to satelite, and where does it stores the data in fprime (database etc.,)? and any documentation on integrating with cosmos etc.,
![image](https://user-images.githubusercontent.com/1790810/161828511-87858309-8a5f-455a-a528-316c85ad0723.png)

**Reply by LeStarch** [reply]
This is typically an error in the GDS itself.  Are you seeing any output on the console where you ran `fprime-gds`?

**Comment by nchanilk** [comment]
console is clean, i don't see any error while running or after accessing the UI. I enabled nginx webserver aswell, but still it's the same problem. Please find below console screen shot
![image](https://user-images.githubusercontent.com/1790810/161837456-c81c956a-2761-4e9b-bf78-d19bb3da7c7a.png)

**Reply by LeStarch** [reply]
Can you look at the  text of the response  up above in the browser dev-tools? See what the actual reply was from the server was?

It says status 400, but I am wondering what data came back with that status 400.

**Comment by nchanilk** [comment]
PFA
getting the" {"message": "The browser (or proxy) sent a request that this server could not understand."}
![image](https://user-images.githubusercontent.com/1790810/161839044-7997a7d6-9002-4d69-b14a-cf6e526d0624.png)
![image](https://user-images.githubusercontent.com/1790810/161839135-d7a9bc0e-be0b-403f-a963-f714c01cd9ef.png)

**Reply by LeStarch** [reply]
It looks right.  You said you setup `nginx`.  Is it configured correctly as a proxy?  What happens if you access the GDS directly (localhost:5000)?  Below is how I've seen proxies done before in nginx.

```
    location / {
        rewrite ^/(.*)$ /$1 break;
        proxy_pass http://127.0.0.1:5000/;

        proxy_set_header    Host               $host;
        proxy_set_header    X-Real-IP          $remote_addr;
        proxy_set_header    X-Forwarded-For    $proxy_add_x_forwarded_for;
    }
```

**Comment by nchanilk** [comment]
The initial error which i posted was without nginx, ( the error is same with /with out nginix)
I've replaced the above given nginix configuration, but still same issue.
for your reference below is my nginx code
`location / {
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-Proto https;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-NginX-Proxy true;
        proxy_pass http://localhost:5000/;
        proxy_set_header Host $http_host;
        proxy_cache_bypass $http_upgrade;
        proxy_redirect off;
        add_header Access-Control-Allow-Origin *;
     }`

**Reply by LeStarch** [reply]
Have you tried a force reload of the page?  e.g. CTRL-SHIFT-R?  This would force a reload in case something got out of sync.

**Reply by LeStarch** [reply]
At this point I cannot see what is corrupting the request, but something seems to be scrambling it such that the flask server cannot understand it. This could be caused by a bad proxy, an out-of-sync version of the webpage.  However, I am not sure what else might cause this.  The oddity is that some endpoints (files, logdata) still work but not others.

**Reply by LeStarch** [reply]
Just to check, you installed these tools via pip?  e.g. pip install fprime-tools fprime-gds?

**Comment by nchanilk** [comment]
Here are the commands executed
![image](https://user-images.githubusercontent.com/1790810/161848226-b0ea27fd-e1ce-49e6-98ef-5155fbb71b3f.png)
![image](https://user-images.githubusercontent.com/1790810/161848321-20c7b639-ad67-4232-a759-68471d6102fc.png)
![image](https://user-images.githubusercontent.com/1790810/161848884-4d3b9615-e09b-43af-9f93-5bb49b9cfd53.png)
I remember having few issues with the flask and I've to downgrade. Also I'm using python 3.8.10, Do you think this may be an issue? 
Please let me know the python and flask versions, I'll try do a clean set up.
Please let me know the correct versions, I'll kill this vm and spin up freshly and ensure not to have any conflicts

**Reply by LeStarch** [reply]
Here are the versions I have:

```
flask --version
Python 3.8.2
Flask 1.1.2
Werkzeug 2.0.1
```

**Reply by LeStarch** [reply]
I don't think you got your flask installed properly (I see 2.1.1 above).

**Comment by nchanilk** [comment]
Sounds good, thank you. I'll try to clean up everything and set up based on the above configuration.
Thanks, @LeStarch  for all your help. I'll keep you posted by today/tomorrow on how it goes.

**Comment by nchanilk** [comment]
Yea, I remember now, if I use Flask==1.1.2 then I ran into multiple improt issues like below, because of that I upgraded to 2.1.1
![image](https://user-images.githubusercontent.com/1790810/161855151-6523dedc-bcf4-4268-aeef-dcd7d25eac80.png)

**Comment by nchanilk** [comment]
Thanks, @LeStarch for your help. I'm not getting the console errors now after fixing my versions and did clean set up.
again, thanks a lot for your help.
I'm still learning the stuff, and I need to understand more on this, can you please guide me to the right channel?
1. Where is the data saving, is it saving in a database? 
2. What database are we using for this?
3. I want to to send the sequence commands and receive telemetry data from the satelite, how to do that? where are the connections exists? and what/where need to be modified?
4. How can we feed the telemetry data from satelite to the UI through the fprime?
5. We're looking for cosmos5 integration, https://github.com/fprime-community/fprime-cosmos-integration/blob/master/README.md, any guidance would be helpful

I might be asking dumb questions, apologies for that. Any constructive guidance would be a great help.

Thanks,
Anil

**Comment by nchanilk** [comment]
@LeStarch   When you get a chance, could you please help with the above?

**Reply by LeStarch** [reply]
Sure!  The first point to know is that this GDS was built for in-lab or on-ground testing.  Most projects are told to use a specific ground system and thus use this GDS for testing until their project provides the desired ground system.

I provided an in-depth explanation here: https://github.com/nasa/fprime/discussions/1269  that seems to answer many of your questions, but I can provide short answers here.

1. We save data in RAM.  One can integrate with a database of choice by writing a adapter to push/pull from that database as discussed above.
2. No database is specifically supported.
3. For the hardware connection, you'd need to write another adapter for your comm to data stream.  https://github.com/fprime-community/fprime-gds/blob/devel/src/fprime_gds/common/communication/adapters/base.py is the base class and we have IP and UART adapters for in-lab testing.  Whatever your data stream might be (radio, communications provider, etc) you can adapt that to bring it in to the ground station.
4. See item 3
5. We don't have a lot of support for COSMOS integration.  It is provided as a reference for projects that would like to try and go that route. There are a few more notes on COSMOS usage in the above discussion #1269

**Comment by nchanilk** [comment]
Thanks, @LeStarch. I'll go through the documentation and your answers.
